File: 2024-mk2/arrays/best-time-to-buy-and-sell-stock.py
"""
You are given an integer array prices where prices[i] is the price of NeetCoin on the ith day.

You may choose a single day to buy one NeetCoin and choose a different day in the future to sell it.

Return the maximum profit you can achieve. You may choose to not make any transactions, in which case the profit would be 0.
"""
from typing import List
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        res = 0
        l,r = 0, 1

        while r < len(prices):
            if prices[r] < prices[l]:
                l = r
                r += 1
            else:
                res = max((prices[r]-prices[l]), res)
                r+=1
        return res

sol = Solution()
print(sol.maxProfit([10,1,5,6,7,1]))

File: 2024-mk2/two-pointers/best-time-to-buy-and-sell-stock.go
package main

import "fmt"

type DataPt struct {
  buy int;
  sell int;
  profit int;
}

func maxProfit(prices []int, dp *DataPt) int {
  l, r := 0, 1

  for r < len(prices) {
    if prices[r] < prices[l] {
      l = r
      r++
      continue
    }
    tmp := prices[r] - prices[l]
    if tmp > dp.profit {
      dp.profit = tmp
      dp.buy = l
      dp.sell = r
    }
    r++
  }
  fmt.Println(dp.buy)
  fmt.Println(dp.sell)

  return dp.profit
}

func main() {
  dp := &DataPt{0, 1, 0}
  prices := []int{7, 1, 5, 3, 6, 4}
  fmt.Println(maxProfit(prices, dp))

  dp = &DataPt{0, 1, 0}
  prices2 := []int{7, 6, 4, 3, 1}
  fmt.Println(maxProfit(prices2, dp))

}
